### ISSUE TITLE
Discussion: Use of draft pull request feature

### ISSUE THREAD

**Issue by yhtMinceraft1010X** [nus-cs2103-AY2021S2/forum#236]
I can think of 2 approaches to open pull requests:

1. Open only when ready (New feature/change fully implemented in branch + tests and checkstyle cleared)
2. Open as a **draft pull request** once you make a commit (Even if not fully implemented)

Personally, I think option 2 should be considered as it allows for teammates to see your progress as you implement whatever change you are assigned (so if you get stuck anywhere, someone might be able to offer advice early on). A reviewer could point out changes to consider implementing while developing a feature rather than wait until the assignee thinks the feature is finished before asking the assignee to go back and edit the code.

The only drawback I can think of is that there would be a lot more CI reports (including failing ones) generated.

What does everyone else think?

**Comment by tjtanjin** [comment]
I think it depends on the scale of the feature being worked on as well as the approach chosen to work on it. If a small feature was being worked on, then draft requests may not be necessary as pull requests completing smaller features should in theory be more manageable.

On the other hand, if a feature is big enough to warrant discussion/help, then perhaps that justifies the need for a draft pull request although I am under the impression that it is usually best to break down a huge feature into smaller tasks which then brings me back to my first point.

All that said I think it is possible to use draft pull requests without constantly failing CI if coding standards are adhered to and the affected test cases temporarily commented out (probably not the ideal practice but a solution nonetheless). Githooks may also be setup locally to prevent instances of commits done without first passing checkstyle and test cases locally (I personally think this helps a lot with reducing unnecessary "fix checkstyle commits").

**Comment by tlylt** [comment]
After using it myself and also with @tjtanjin, I think draft PR is a good way to inform your teammates that you are working on certain tasks/features and this might save some typing (as an automatic email will be sent to notify everyone about the PR).
One possible drawback is that whenever a commit is updated in the draft PR, an email will be sent to your teammates as well. This might be deemed as unwarranted/unnecessary noise to some.

**Comment by ong6** [comment]
I agree with @tjtanjin and @tlylt, after seeing @tjtanjin use it for our team, I think the draft PR is pretty good.

A possible alternative could be to add a tag with a "still working on this" on a normal PR, in case you want to let your teammates know. (To prevent getting spammed with emails)

**Comment by JulietTeoh** [comment]
I think that the use of the draft pull request feature can really be helpful for my team because it could reduce instances where we would have to review large chunks of code in one go when there is a large pull request. Thanks for informing everyone on such an enlightening feature!
